### PR TITLE
feat:chain:splitstore auto prune

### PR DESCRIPTION
**Pull request by ZenGround0**
## Proposed Changes

Auto prune of coldstore
1) Add logic to trigger chain prune on head events
2) Add config options to support this

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

File: blockstore/splitstore/splitstore.go [hunk 1]
```diff
@@ -676,6 +695,23 @@ func (s *SplitStore) Start(chain ChainAccessor, us stmgr.UpgradeSchedule) error
...
+	case nil:
+		s.pruneEpoch = bytesToEpoch(bs)
+	case dstore.ErrNotFound:
+		if curTs == nil {
+			//this can happen in some tests
+			break
+		}
+		if err := s.setPruneEpoch(curTs.Height()); err != nil {
+			return xerrors.Errorf("error saving prune epoch: %w", err)
+		}
```
Comment (magik6k) [inline blockstore/splitstore/splitstore.go:710]: Should also set s.pruneEpoch to current height here.

File: blockstore/splitstore/splitstore_compact.go [hunk 2]
```diff
@@ -81,6 +83,9 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
 	curTs := apply[len(apply)-1]
 	epoch := curTs.Height()
 
+	log.Warnf("Head epoch %d", epoch)
+	fmt.Printf("auto prune enabled: %t, gap %d from pruneEpoch %d, compaction index %d\n", s.cfg.EnableColdStoreAutoPrune, epoch-s.pruneEpoch, s.pruneEpoch, s.compactionIndex)
```
Comment (magik6k) [inline blockstore/splitstore/splitstore_compact.go:87]: debug log

File: blockstore/splitstore/splitstore_compact.go [hunk 3]
```diff
@@ -130,6 +137,40 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
...
+					return depth <= int64(CompactionBoundary)+s.cfg.ColdStoreRetention*int64(build.Finality)
+				}
+			case s.cfg.ColdStoreRetention < 0:
+				retainP = func(_ int64) bool { return true }
+			default:
+				retainP = func(depth int64) bool {
+					return depth <= int64(CompactionBoundary)
+				}
+			}
+			movingGC := s.cfg.HotStoreFullGCFrequency > 0 && s.pruneIndex%int64(s.cfg.ColdStoreFullGCFrequency) == 0
```
Comment (magik6k) [inline blockstore/splitstore/splitstore_compact.go:164]: Shouldn't this be
```suggestion
			movingGC := s.cfg.ColdStoreFullGCFrequency > 0 && s.pruneIndex%int64(s.cfg.ColdStoreFullGCFrequency) == 0
```

File: blockstore/splitstore/splitstore.go (hunk 1 above)
Comment (ZenGround0) [reply blockstore/splitstore/splitstore.go:710]: This is done inside setPruneEpoch along with the datastore write

File: blockstore/splitstore/splitstore_compact.go [hunk 4]
```diff
@@ -81,6 +82,8 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
 	curTs := apply[len(apply)-1]
 	epoch := curTs.Height()
 
+	log.Warnf("Head epoch %d", epoch)
```
Comment (magik6k) [inline blockstore/splitstore/splitstore_compact.go:85]: Probably don't want this log

**Review by magik6k — APPROVED** [review]

**Pull request merged**